helper/schema/resource.go

```diff
@@ -174,6 +183,51 @@ type Resource struct {
 	Timeouts *ResourceTimeout
 }
 
+// Contextify upgrades all set non-context funcs (CRUD, Exists, CustomizeDiff) to the
+// context aware counterparts.
+func (r *Resource) Contextify() *Resource {
```

Comment:
I think this method is a bit leaky. It should probably not be public as nobody outside should need to call it, and it has a lot of branching complexity.

You'd probably be better off just creating like an `r.create`, `r.read` or something (maybe `readInternal`), etc. Then your calls elsewhere are still a single func without their own branching logic but you don't have a public func that mutates schema.

Each function then could apply the timeout and bonus, they could easily `defer` the cancel in a tight scope.

```go
func (r *Resource) deleteInternal(ctx context.Context, ...) ... {
  if r.DeleteContext == nil {
    return r.Delete(...)
  }

  ctx, cancel := context.WithTimeout(ctx, data.Timeout(TimeoutDelete))
  defer cancel()
  return r.DeleteContext(ctx, ...)
}
```

Alternatively since the CRUD func signatures are so similar you could have something like:

```go
func shimContext(ctx context.Context, ctxFunc func(context.Context, *ResourceData, interface{}) error, legacyFunc func(*ResourceData, interface{}) error, timeout, d *ResourceData, meta interface{}) error {
  if ctxFunc == nil && legacyFunc == nil {
    // error at least one needs to be set
  }
 
  if ctxFunc == nil {
    return legacyFunc(...)
  }
  
  ctx, cancel := context.WithTimeout(ctx, timeout)
  defer cancel()
  return ctxFunc(...)
}

err := shimContext(ctx, r.CreateContext, r.Create, TimeoutCreate, d, meta)
...
```

Either way though its more functional and not mutating and the branching and responsibilities are kept smaller.

Reply:
I agree it felt awkward as a public method. I do like the tighter `readInternal` suggestion, this works well for the CRUD functions but the `CustomizeDiffFunc` gets passed around by reference in a few spots (the codepaths there are a bit messier). I won't have the same opportunity to conditionally evaluate if a "context aware CDF" is set, it was easier to simply switch to the new function signature and upgrade through `contextify()`. I have now privatized this method and during runtime it is only called once, during `Resource.InternalValidate()` this also makes validation easier as I don't need double checks (`if r.Create || r.CreateContext`).

Reply:
the others calls to `contextify()` are only called in tests, this was to avoid rewriting a zillion function signatures.